File: spring-web-java/src/main/java/spring/web/java/dto/response/ApiResponse.java
package spring.web.java.dto.response;

import com.fasterxml.jackson.annotation.JsonInclude;
import spring.web.java.constant.ApiResponseCode;


public record ApiResponse<T>(
        String code,
        String message,
        @JsonInclude(JsonInclude.Include.NON_NULL)
        T payload
) {
    public ApiResponse() {
        this(ApiResponseCode.SUCCESS.name(), ApiResponseCode.SUCCESS.getDescription(), null);
    }

    public ApiResponse(T payload) {
        this(ApiResponseCode.SUCCESS.name(), ApiResponseCode.SUCCESS.getDescription(), payload);
    }

    public ApiResponse(ApiResponseCode apiResponseCode) {
        this(apiResponseCode.name(), apiResponseCode.getDescription(), null);
    }

    public ApiResponse(ApiResponseCode apiResponseCode, T payload) {
        this(apiResponseCode.name(), apiResponseCode.getDescription(), payload);
    }
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/dto/response/ApiResponse.kt
package spring.web.kotlin.dto.response

import com.fasterxml.jackson.annotation.JsonInclude
import spring.web.kotlin.constant.ApiResponseCode

data class ApiResponse<T>(
    val code: String,
    val message: String,
    @JsonInclude(JsonInclude.Include.NON_NULL)
    val payload: T?
) {
    constructor() : this(
        code = ApiResponseCode.SUCCESS.name,
        message = ApiResponseCode.SUCCESS.description,
        payload = null
    )
    constructor(payload: T?) : this(
        code = ApiResponseCode.SUCCESS.name,
        message = ApiResponseCode.SUCCESS.description,
        payload = payload
    )
    constructor(apiResponseCode: ApiResponseCode) : this(
        code = apiResponseCode.name,
        message = apiResponseCode.description,
        payload = null
    )
    constructor(apiResponseCode: ApiResponseCode, payload: T?) : this(
        code = apiResponseCode.name,
        message = apiResponseCode.description,
        payload = payload
    )
}
